.github/workflows/gpu_test.yml

```diff
@@ -0,0 +1,48 @@
+# This workflow will install Python dependencies, run tests and lint with a variety of Python versions
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: GPU Test V3
+
+on:
+  push:
+    branches: [ v3 ]
+  pull_request:
+    branches: [ v3 ]
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  test:
+    name: py=${{ matrix.python-version }}, np=${{ matrix.numpy-version }}, deps=${{ matrix.dependency-set }}
+
+    runs-on: gpu-runner
+    strategy:
+      matrix:
+        python-version: ['3.10', '3.11', '3.12']
+        numpy-version: ['1.24', '1.26', '2.0']
+        dependency-set: ["minimal", "optional"]
+
+    steps:
+    - uses: actions/checkout@v4
+    - name: GPU check
+      run: |
+        nvidia-smi
+    - name: Set up Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: ${{ matrix.python-version }}
+        cache: 'pip'
+    - name: Install Hatch and CuPy
+      run: |
+        python -m pip install --upgrade pip 
+        pip install hatch cupy-cuda12x
```

Comment:
Manually adding this here for now. Should make this part of the hatch workflow later.

Reply:
Yeah there's probably a better way to handle this. Happy to chat when this is ready for discussion

Reply:
This is the only thing remaining for this PR, so any solutions here would be great!

Reply:
Great, did an initial pass below

We should be able to use [the self-referential extra with hatch]( https://hatch.pypa.io/1.12/config/dependency/#self-referential ) as well